Mathlib/MeasureTheory/SetAlgebra.lean

```diff
@@ -0,0 +1,237 @@
+  empty_mem := h𝒜.empty_mem
+  union_mem := h𝒜.union_mem
+  diff_mem := fun _ _ ↦ h𝒜.diff_mem
+
+/-- An algebra of sets is closed by finite unions. -/
+theorem biUnion_mem {ι : Type*} (h𝒜 : IsSetAlgebra 𝒜) {s : ι → Set α} (S : Finset ι)
+    (hs : ∀ i ∈ S, s i ∈ 𝒜) : ⋃ i ∈ S, s i ∈ 𝒜 := h𝒜.isSetRing.biUnion_mem S hs
+
+/-- An algebra of sets is closed by finite intersections. -/
+theorem biInter_mem {ι : Type*} (h𝒜 : IsSetAlgebra 𝒜) {s : ι → Set α} (S : Finset ι)
+    (hs : ∀ i ∈ S, s i ∈ 𝒜) : ⋂ i ∈ S, s i ∈ 𝒜 := by
+  by_cases h : S = ∅
+  · rw [h, ← Finset.set_biInter_coe, Finset.coe_empty, biInter_empty]
+    exact h𝒜.univ_mem
+  · rw [← ne_eq, ← Finset.nonempty_iff_ne_empty] at h
+    exact h𝒜.isSetRing.biInter_mem S h hs
+
+end IsSetAlgebra
+
+section generateSetAlgebra
+
+/-! ### Definition and properties of the algebra of sets generated by some family -/
+
+/-- `generateSetAlgebra 𝒜` is the smallest algebra of sets containing `𝒜`. -/
+inductive generateSetAlgebra {α : Type*} (𝒜 : Set (Set α)) : Set (Set α)
+  | base (s : Set α) (s_mem : s ∈ 𝒜) : generateSetAlgebra 𝒜 s
+  | empty : generateSetAlgebra 𝒜 ∅
+  | compl (s : Set α) (hs : generateSetAlgebra 𝒜 s) : generateSetAlgebra 𝒜 sᶜ
+  | union (s t : Set α) (hs : generateSetAlgebra 𝒜 s) (ht : generateSetAlgebra 𝒜 t) :
+      generateSetAlgebra 𝒜 (s ∪ t)
+
+/-- The algebra of sets generated by a family of sets is an algebra of sets. -/
+theorem isSetAlgebra_generateSetAlgebra :
+    IsSetAlgebra (generateSetAlgebra 𝒜) where
+  empty_mem := generateSetAlgebra.empty
+  compl_mem := fun _ hs ↦ generateSetAlgebra.compl _ hs
+  union_mem := fun _ _ hs ht ↦ generateSetAlgebra.union _ _ hs ht
+
+/-- The algebra of sets generated by `𝒜` contains `𝒜`. -/
+theorem self_subset_generateSetAlgebra : 𝒜 ⊆ generateSetAlgebra 𝒜 :=
+  fun _ ↦ generateSetAlgebra.base _
+
+/-- The measurable space generated by a family of sets `𝒜` is the same as the one generated
+by the algebra of sets generated by `𝒜`. -/
+@[simp]
+theorem generateFrom_generateSetAlgebra_eq :
+    generateFrom (generateSetAlgebra 𝒜) = generateFrom 𝒜 := by
+  refine le_antisymm (fun s ms ↦ ?_) (generateFrom_mono self_subset_generateSetAlgebra)
+  refine @generateFrom_induction _ _ (generateSetAlgebra 𝒜) (fun t ht ↦ ?_)
+    (@MeasurableSet.empty _ (generateFrom 𝒜))
+    (fun t ↦ MeasurableSet.compl)
+    (fun f hf ↦ MeasurableSet.iUnion hf)
+    s ms
+  induction ht with
+  | base u u_mem => exact measurableSet_generateFrom u_mem
+  | empty => exact @MeasurableSet.empty _ (generateFrom 𝒜)
+  | compl u _ mu => exact mu.compl
+  | union u v _ _ mu mv  => exact MeasurableSet.union mu mv
+
+/-- If a family of sets `𝒜` is contained in `ℬ`, then the algebra of sets generated by `𝒜`
+is contained in the one generated by `ℬ`. -/
+theorem generateSetAlgebra_mono {ℬ : Set (Set α)} (h : 𝒜 ⊆ ℬ) :
+    generateSetAlgebra 𝒜 ⊆ generateSetAlgebra ℬ := by
+  intro s hs
+  induction hs with
+  | @base t t_mem => exact self_subset_generateSetAlgebra (h t_mem)
```

Comment:
You're still using the `@` version here and below (several times). Is that on purpose?

Reply:
Sorry I forgot them